src/Dialect/ONNX/ONNXOpsHelper.cpp

```diff
@@ -42,13 +44,84 @@ Value OnnxBuilder::matmul(Type Y, Value A, Value B) const {
   return b.create<ONNXMatMulOp>(loc, Y, A, B);
 }
 
-Value OnnxBuilder::reshape(Type outputType, Value input, Value shape) const {
-  return b.create<ONNXReshapeOp>(loc, outputType, input, shape);
+Value OnnxBuilder::reshape(
+    const Value input, const ArrayRef<DimIndexExpr> shapeDims) const {
+  assert(!shapeDims.empty() && "Shape dimensions should not be empty");
+
+  ShapedType inputType = input.getType().cast<ShapedType>();
+  Type elementType = inputType.getElementType();
+
+  // If the output dimensions are all literals the 'onnx/Reshape' operation
+  // can take the new shape via an 'onnx.Constant'.
+  if (llvm::all_of(
+          shapeDims, [](const DimIndexExpr &dim) { return dim.isLiteral(); })) {
+    SmallVector<int64_t, 6> shape;
+    for (const IndexExpr &dim : shapeDims)
+      shape.push_back(dim.getLiteral());
+
+    auto constantOp =
+        getONNXConstantOpFromDenseAttr(b, loc, b.getI64TensorAttr(shape));
+
+    Value reshapeRes = b.create<ONNXReshapeOp>(
+        loc, MemRefType::get(shape, elementType), input, constantOp);
+
+    return reshapeRes;
+  }
+
+  MemRefBuilder memRefBuilder(b, loc);
+  KrnlBuilder krnlBuilder(memRefBuilder);
+
+  // When the output dimensions aren't all literals we need to generate code
+  // to compute the shape. Allocate a buffer and store the putput dimension
+  // into it.
+  IndexType indexTy = b.getIndexType();
+  int64_t length = shapeDims.size();
+  memref::AllocOp alloc =
+      memRefBuilder.alignedAlloc(MemRefType::get({length}, indexTy), 16);
+
+  for (int64_t i = 0; i < length; ++i) {
+    Value index = emitConstantOp(b, loc, indexTy, i);
+    Value data = shapeDims[i].getValue();
+    krnlBuilder.store(data, alloc, index);
+  }
+
+  // Now create the 'onnx.Reshape' operation. Because the shape is not a
+  // compile time constant it is effectively unknown.
+  SmallVector<int64_t> shape(length, -1);
+  Value reshapeRes = b.create<ONNXReshapeOp>(
+      loc, MemRefType::get(shape, elementType), input, alloc);
```

Comment:
Instead of preparing `shape(length, -1)` and calling MemRefCast, could we replace `shape` by `castOutputShape` here? So there is no need to call MemRefCast.

Reply:
I gave that a try before and got the following error: 

```
 TEST_CASE_BY_USER=test_depthtospace_example_cpu  TEST_DYNAMIC=true /usr/local/bin/python3.9 /Users/etiotto@ca.ibm.com/Source/onnx/onnx-mlir/build/test/backend/Debug/test.py 
 ...
 error: expected result type with size = -1 instead of 2 in dim = 1
// -----// IR Dump After (anonymous namespace)::FrontendToKrnlLoweringPass Failed //----- //
```

I believe the reason for the error is that the new shape is a variable and so the Reshape output type cannot have any constant extents. It expects the output type to be memref<?x?x?x?x?x?xf32> rather than memref<?x2x2x?x?x?xf32>:
```
%14 = memref.alloc() {alignment = 16 : i64} : memref<6xindex>
. . .
krnl.store %10, %14[%c5] : memref<6xindex>
%15 = "onnx.Reshape"(%arg0, %14) : (memref<?x?x?x?xf32>, memref<6xindex>) -> memref<?x2x2x?x?x?xf32>
```
 
That is the reason I ended up injecting a cast aster the Reshape.

Reply:
I see. Thanks for checking!